lib/grist-api.ts

```diff
@@ -0,0 +1,284 @@
+    const filters = options.filters;
+
+    // Maps unique keys to Grist rows
+    const gristRows = new Map<string, IRecord>();
+    for (const oldRec of await this.fetchTable(tableName, filters)) {
+      const key = JSON.stringify(keyColIds.map((colId) => oldRec[colId]));
+      gristRows.set(key, oldRec);
+    }
+
+    const updateList: IRecord[] = [];
+    const addList: IRecord[] = [];
+    let dataCount = 0;
+    let filteredOut = 0;
+    for (const newRec of records) {
+      if (filters && keyColIds.some((colId) => filters[colId] && !filters[colId].includes(newRec[colId]))) {
+        filteredOut += 1;
+        continue;
+      }
+      dataCount += 1;
+      const key = JSON.stringify(keyColIds.map((colId) => newRec[colId]));
+      const oldRec = gristRows.get(key);
+      if (oldRec) {
+        const changedKeys = Object.keys(newRec).filter((colId) => newRec[colId] !== oldRec[colId]);
+        if (changedKeys.length > 0) {
+          debug("syncTable %s: #%s %s needs updates", tableName, oldRec.id, key,
+            changedKeys.map((colId) => [colId, oldRec[colId], newRec[colId]]));
+          const update: IRecord = pick(newRec, changedKeys);
+          update.id = oldRec.id;
+          updateList.push(update);
+        }
+      } else {
+        debug("syncTable %s: %s not in grist", tableName, key);
+        addList.push(newRec);
+      }
+    }
+
+    debug("syncTable %s (%s) with %s records (%s filtered out): %s updates, %s new",
+      tableName, gristRows.size, dataCount, filteredOut, updateList.length, addList.length);
+    await this.updateRecords(tableName, updateList);
+    await this.addRecords(tableName, addList);
+  }
+
+  /**
+   * Low-level interface to make a REST call.
+   */
+  private async _call(url: string, jsonData?: object, method?: Method) {
+    const fullUrl = `${this._server}/api/docs/${this._docId}/${url}`;
+
+    method = method || (jsonData ? 'POST' : 'GET');
+    if (this._dryrun && method.toUpperCase() !== 'GET') {
+      debug("DRYRUN NOT sending %s request to %s", method, fullUrl);
+      return;
+    }
+    if (!this._apiKey) {
+      // If key is missing, get it on first use (possibly from a file), since the constructor can't be async.
+      this._apiKey = await getAPIKey();
+    }
+    debug("Sending %s request to %s", method, fullUrl);
+    try {
+      const request = {
+        url: fullUrl,
+        method,
+        data: jsonData,
+        headers: {
+          'Authorization': `Bearer ${await this._apiKey}`,
+          'Content-Type': 'application/json',
+          'Accept': 'application/json',
+        },
+      };
+      debugReq("Request", request);
+      const response = await axios.request(request);
+      return response.data;
+    } catch (err) {
+      // If the error has {"error": ...} content, use that for the error message.
+      const errorObj = err.response ? err.response.data : null;
+      if (typeof errorObj === 'object' && errorObj && errorObj.error) {
```

Comment:
This is nitpicking : if `errObj` is of type 'object' it's then necessarily truthy it is enough to write: `if (typeof errorObj === 'object' && errorObj.error) {...}`.

Reply:
I am trying to protect against `null` -- and `typeof null` is `'object'`

Reply:
Oups, I keep forgetting about this one.